### ISSUE TITLE
`fn_addr_eq` docs suggest incorrect reasoning

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust#160202]
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/issues/160202)*

The [docs](https://doc.rust-lang.org/nightly/std/ptr/fn.fn_addr_eq.html) for `fn_addr_eq` say

> Despite these false positives and false negatives, this comparison can still be useful. Specifically, if
> -  T is the same type as U, T is a [subtype](https://doc.rust-lang.org/reference/subtyping.html) of U, or U is a [subtype](https://doc.rust-lang.org/reference/subtyping.html) of T, and
> -  ptr::fn_addr_eq(f, g) returns true,
>
> then calling f and calling g will be equivalent.

This is not correct, I think.
Consider this example:
```rust
unsafe fn f(x: i32) -> i32 { x }
unsafe fn g(x: i32) -> i32 { std::hint::assert_unchecked(x != 0); x }

fn main() {
  let fptr = f as unsafe fn(i32) -> i32;
  let gptr = g as unsafe fn(i32) -> i32;
  if std::ptr::fn_addr_eq(fptr, gptr) {
    // SAFETY: Calling `fptr(0)` here would be allowed, and according to the `fn_addr_eq`
    // docs, inside this `if` calling `gptr` is equivalent to calling `fptr`.
    unsafe { gptr(0) };
  }
}
```
The compiler may now do the following:
- Realize that `gptr` points to `g`, and use that to inline the call to `g`.
- Optimize the `then` branch to `std::hint::assert_unchecked(false)`, and then replace it by arbitrary code (e.g. a trap).
- Later during compilation, merge `f` and `g` into one function since they compile to the same assembly.

Now if we run this program we hit the trap, i.e., we have exhibited UB.
(See https://github.com/rust-lang/unsafe-code-guidelines/issues/589 for a wider discussion of the problem. When discussing this we were not aware that our libs docs actually make any claims in this regard.)

Cc @rust-lang/opsem @rust-lang/libs-api

**Comment by RalfJung** [comment]
FWIW this example also shows that it would be unsound for LLVM to replace one function pointer by another even if they are equal. Which in turn implies function pointers must have provenance. Cc @nikic

**Comment by theemathas** [comment]
Another alternative for LLVM would be to disable devirtualization as soon as we "observe" a function pointer's address.

**Comment by RalfJung** [comment]
That would entirely kill devirtualization since LLVM will basically ~never be able to prove that we did *not* observe a function's address.

Consider that the `gptr` call can be arbitrarily deep inside other functions and whatnot. Our SAFETY reasoning is allowed to span across functions.

**Comment by ds84182** [comment]
During compilation LLVM makes no assumptions about function addresses since the final address is decided by the static or dynamic linker. So LLVM cannot optimize that function address equality check to true or false.

The docs for fn_addr_eq should be changed to remove the equivalence guarantee.

I don't see how LLVM replacing function pointers leads to unsoundness _after_ the docs are changed.

**Comment by RalfJung** [comment]
>  So LLVM cannot optimize that function address equality check to true or false.

We agree on that. No part of my argument involved optimizing the equality check so I am not sure why you bring that up.


> I don't see how LLVM replacing function pointers leads to unsoundness after the docs are changed.

Indeed, and nobody claimed otherwise as far as I know.

**Comment by RalfJung** [comment]
> The docs for fn_addr_eq should be changed to remove the equivalence guarantee.

So the docs will just say "this returns a `bool`, you cannot do anything useful with it"...

I don't have a better idea either. :/

**Comment by asquared31415** [comment]
The concerns that comparing function pointers is not very useful were brought up in https://github.com/rust-lang/rust/pull/118833 which added a lint against comparing function pointers (because it's inherently a sketchy operation).  Then `fn_addr_eq` was spawned off that lint, for the use case of "i want to do it anyway", but the concerns about what it means didn't come up in the [ACP](https://github.com/rust-lang/libs-team/issues/323), [implementation](https://github.com/rust-lang/rust/pull/129323), or [tracking issue](https://github.com/rust-lang/rust/issues/129322), and were never addressed.

**Comment by RalfJung** [comment]
Nominated for t-libs to get their take on this.

On the opsem side, my understanding is that the way https://github.com/llvm/llvm-project/issues/159565 has been fixed is by acknowledging that function pointers have provenance.  The current docs on `fn_addr_eq` essentially state that function pointers do not have provenance. That's just fundamentally in conflict with reality. And this isn't really LLVM's fault; once you allow function deduplication based on the *assembly* code being the same, you pretty much have to use provenance to disambiguate which function is actually meant. I don't know a different semantics they could use for `unnamed_addr` functions.

**Comment by Darksonn** [comment]
Should we just fix the docs and add your example as a warning?

**Comment by RalfJung** [comment]
What should the docs say? "Don't call this, it doesn't give you any useful information"?

**Comment by Urgau** [comment]
T-lang should probably be involved, [they are the one who originally asked for it](https://github.com/rust-lang/rust/pull/118833#issuecomment-1875784492).

**Comment by asquared31415** [comment]
Fundamentally, this function is trying to be a "blessed" alternative for `#[allow(unpredictable_function_pointer_comparisons, reason="i dont care about correctness")]`.  It should never have existed.

**Comment by RalfJung** [comment]
> T-lang should probably be involved, [they are the one who originally asked for it](https://github.com/rust-lang/rust/pull/118833#issuecomment-1875784492).

They asked for a function with the same semantics as `==` to have a good place to put docs, but were not involved in specifying it AFAIK? But yeah we can nominate for them as well.

**Comment by nia-e** [comment]
Per today's libs meeting, we agree to change the docs to delete the bit about calls being equivalent (worth mentioning this in docs so somebody doesn't trip up on it, tbf). There are probably cases where this is usable where it's just a behavioural correctness thing (e.g. to check if some callback is set) but this should probably trigger a lint. Feel free to request an FCP on such a PR!

**Comment by traviscross** [comment]
> T-lang should probably be involved, [they are the one who originally asked for it](https://github.com/rust-lang/rust/pull/118833#issuecomment-1875784492).

cc @rust-lang/lang

**Comment by RalfJung** [comment]
Here we go: https://github.com/rust-lang/rust/pull/162140.
Feedback on the wording is welcome!

**Comment by asquared31415** [comment]
I'm unsure if equality is *ever* correct, because functions in Rust don't have an identity.  Equality is not a meaningful operation under those conditions.  We effectively have a type that implements `Eq` but satisfies **none** of the requirements: symmetry, transitivity, and reflexivity are all broken, and the relationship between `==` and `!=` may also be broken.

**Comment by RalfJung** [comment]
> symmetry, transitivity, and reflexivity are all broken, 

That's not quite true... the following assertions can never fail:
```rust
fn check(a: fn(), b: fn(), c: fn()) {
  assert!(a == a && b == b && c == c);
  if a == b { assert!(b == a); }
  if a == b && b == c { assert!(a == c); }
}
```
IOW, the domain of `fn()` pointers at runtime does have a genuine equivalence relation.

However, the mapping from function items to `fn()` pointers is non-deterministic and non-obvious, and that's where the problems come from.

**Comment by asquared31415** [comment]
Don't optimizations allow the compiler to turn a `fn()` pointer into its function item and then potentially back into a *different* `fn()` pointer?

**Comment by RalfJung** [comment]
No, I don't think so.

**Comment by ChayimFriedman2** [comment]
Isn't it worse than that, i.e. using this function to compare *any* function pointer with a known function is *unsound* and not just incorrect?

Consider:
```rust
mod private {
    fn foo() {
        // SAFETY: This is private and not called.
        unsafe { unreachable_unchecked() }
    }
}
```
And in other file (or crate):
```rust
pub fn bar() {}
```
And assume the compiler deduplicates those functions, and we have a cache of functions (for example, for memoization), as common with function pointer equality, and insert both functions into it but call only `bar()`. This could be turned into a call to `foo()`, which is unsound.

**Comment by RalfJung** [comment]
*Comparing* the function pointers is just incorrect. There's no UB from just calling `fn_addr_eq`.

The unsoundness arises from other code then making conclusions based on that incorrect information. That's always something that can happen with any kind of logical bug which then feeds into data that is relevant for unsafe code.
